debezium-server/debezium-server-redis/src/main/java/io/debezium/server/redis/RedisStreamChangeConsumer.java

```diff
@@ -115,21 +122,61 @@ public void handleBatch(List<ChangeEvent<Object, Object>> records,
             throws InterruptedException {
 
         for (ChangeEvent<Object, Object> record : records) {
-            try {
-
-                LOGGER.trace("Received event '{}'", record);
-
-                String destination = streamNameMapper.map(record.destination());
-                String key = (record.key() != null) ? getString(record.key()) : nullKey;
-                String value = (record.value() != null) ? getString(record.value()) : nullValue;
-                client.xadd(destination, null, Collections.singletonMap(key, value));
-
-                committer.markProcessed(record);
-            }
-            catch (Exception e) {
-                throw new DebeziumException(e);
+            LOGGER.trace("Received event '{}'", record);
+
+            String destination = streamNameMapper.map(record.destination());
+            String key = (record.key() != null) ? getString(record.key()) : nullKey;
+            String value = (record.value() != null) ? getString(record.value()) : nullValue;
+            int currentRetryTime = initialRetryDelay;
+            boolean completedSuccessfully = false;
+
+            // As long as we failed to add the current record to the stream, we should retry if the reason was either a connection error or OOM in Redis.
+            while (!completedSuccessfully) {
+                try {
+                    // Add the record to the destination stream
+                    client.xadd(destination, null, Collections.singletonMap(key, value));
+                    completedSuccessfully = true;
+                }
+                catch (JedisConnectionException jce) {
+                    // Try to reconnect
+                    try {
+                        connect();
+                    }
+                    catch (Exception e) {
+                        LOGGER.error("Can't connect to Redis", e);
+                    }
+                }
+                catch (Exception e) {
+                    // When Redis reaches its max memory limitation, a JedisDataException will be thrown with this message.
+                    // In this case, we will retry adding this record to the stream, assuming some memory will be freed eventually as result
+                    // of evicting elements from the stream by the target DB.
+                    if (e.getMessage().equals("OOM command not allowed when used memory > 'maxmemory'.")) {
```

Comment:
So a server-side OOM is a recoverable situation in Redis? I'm asking because in Java itself you typically should stop the application and restart in case of an OOM. But it might be different for Redis which may benefit from more explicit control over management.

Reply:
Yes, in case some memory will be freed, you'll be able to complete the operation successfully. Our assumption is that the streams are consumed by another application, redis gears, etc. so it might be just a temporarily situation when there's not enough memory to complete the xadd operation.